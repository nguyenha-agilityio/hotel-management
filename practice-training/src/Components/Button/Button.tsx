import { MouseEventHandler } from "react";

export interface ButtonProps {
  label: string;
  backgroundColor?: string;
  size?: "md" | "lg" | "sm";
  handleClick: MouseEventHandler<HTMLButtonElement>;
  borderRadius?: number;
  color?: string;
}

export default function Button({
  label,
  backgroundColor = "red",
  size = "sm",
  handleClick,
  borderRadius = 4,
  color = "white",
}: ButtonProps) {
  let padding = "5px 10px";

  if (size === "lg") padding = "10px 20px";
  else if (size === "md") padding = "7px 15px";

  const style = {
    backgroundColor,
    padding,
    border: "none",
    borderRadius: `${borderRadius}px`,
    color,
    cursor: "pointer",
  };

  return (
    <button onClick={handleClick} style={style}>
      {label}
    </button>
  );
}
