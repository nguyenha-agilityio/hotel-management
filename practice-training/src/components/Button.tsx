import { MouseEventHandler } from "react";

interface ButtonProps {
  label: string;
  backgroundColor: string;
  size: "md" | "lg" | "sm";
  handleClick: MouseEventHandler<HTMLButtonElement>;
}

export default function Button({
  label,
  backgroundColor = "red",
  size = "sm",
  handleClick,
}: ButtonProps) {
  let scale = 1;
  if (size === "sm") scale = 0.75;
  if (size === "lg") scale = 1.5;
  const style = {
    backgroundColor,
    padding: `${scale * 0.5}rem ${scale * 1}rem`,
    border: "none",
  };
  return (
    <button onClick={handleClick} style={style}>
      {label}
    </button>
  );
}
