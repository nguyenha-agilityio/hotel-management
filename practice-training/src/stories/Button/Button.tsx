import { MouseEventHandler } from "react";

export interface ButtonProps {
  label: string;
  backgroundColor: string;
  size: "md" | "lg" | "sm";
  handleClick: MouseEventHandler<HTMLButtonElement>;
  borderRadius: number;
  color: string
}

export default function Button({
  label,
  backgroundColor = "red",
  size = "sm",
  handleClick,
  borderRadius,
  color,
}: ButtonProps) {
  let width = 20;
  if (size === "sm") width = 15;
  if (size === "lg") width = 25;
  const style = {
    backgroundColor,
    padding: `${width * 0.5}px ${width * 1}px`,
    border: "none",
    borderRadius,
    color
  };
  return (
    <button onClick={handleClick} style={style}>
      {label}
    </button>
  );
}
