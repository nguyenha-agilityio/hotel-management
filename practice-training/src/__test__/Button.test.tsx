import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Button from "../Components/Button/Button";
import "@testing-library/jest-dom";

describe("Button Component", () => {
  test("renders correctly with default props", () => {
    const handleClick = jest.fn();
    render(<Button label="Click Me" handleClick={handleClick} />);

    const buttonElement = screen.getByRole("button", { name: /click me/i });
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveStyle("background-color: red");
    expect(buttonElement).toHaveStyle("padding: 5px 10px");
    expect(buttonElement).toHaveStyle("border: 1px");
    expect(buttonElement).toHaveStyle("color: white");
  });

  test("applies custom props", () => {
    const handleClick = jest.fn();
    render(
      <Button
        label="Submit"
        handleClick={handleClick}
        backgroundColor="blue"
        size="lg"
        borderRadius={8}
        color="yellow"
      />,
    );

    const buttonElement = screen.getByRole("button", { name: /submit/i });
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveStyle("background-color: blue");
    expect(buttonElement).toHaveStyle("padding: 10px 20px");
    expect(buttonElement).toHaveStyle("color: yellow");
    expect(buttonElement).toHaveStyle("border-radius: 8px");
    expect(buttonElement).toHaveStyle("border: 2px");
  });

  test("calls handleClick when clicked", () => {
    const handleClick = jest.fn();
    render(<Button label="Click Me" handleClick={handleClick} />);

    const buttonElement = screen.getByRole("button", { name: /click me/i });
    fireEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test("applies border-radius correctly", () => {
    const handleClick = jest.fn();
    render(
      <Button label="Rounded" handleClick={handleClick} borderRadius={12} />,
    );

    const buttonElement = screen.getByRole("button", { name: /rounded/i });
    expect(buttonElement).toHaveStyle("border-radius: 12px");
  });
  test("renders a button with text", () => {
    const handleClick = jest.fn();
    render(<Button label="Click Me" handleClick={handleClick} />);
    const buttonElement = screen.getByText(/click me/i);
    expect(buttonElement).toBeInTheDocument();
  });

  test("matches the snapshot", () => {
    const handleClick = jest.fn();
    const { asFragment } = render(
      <Button label="Snapshot" handleClick={handleClick} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test("applies correct padding for size sm", () => {
    const handleClick = jest.fn();
    render(<Button label="Small" size="sm" handleClick={handleClick} />);

    const buttonElement = screen.getByRole("button", { name: /small/i });
    expect(buttonElement).toHaveStyle("padding: 5px 10px");
  });

  test("applies correct padding for size md", () => {
    const handleClick = jest.fn();
    render(<Button label="Medium" size="md" handleClick={handleClick} />);

    const buttonElement = screen.getByRole("button", { name: /medium/i });
    expect(buttonElement).toHaveStyle("padding: 7px 15px");
  });

  test("applies correct padding for size lg", () => {
    const handleClick = jest.fn();
    render(<Button label="Large" size="lg" handleClick={handleClick} />);

    const buttonElement = screen.getByRole("button", { name: /large/i });
    expect(buttonElement).toHaveStyle("padding: 10px 20px");
  });
});
 