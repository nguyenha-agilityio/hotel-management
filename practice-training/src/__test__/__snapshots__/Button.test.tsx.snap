// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches the snapshot 1`] = `
<DocumentFragment>
  <button
    style="background-color: red; padding: 5px 10px; border-radius: 4px; color: white; cursor: pointer;"
  >
    Snapshot
  </button>
</DocumentFragment>
`;
